src/liballoc/vec.rs

```diff
@@ -1567,40 +1567,84 @@ impl SpecFromElem for u8 {
     }
 }
 
-macro_rules! impl_spec_from_elem {
+impl<T: Clone + IsZero> SpecFromElem for T {
+    #[inline]
+    fn from_elem(elem: T, n: usize) -> Vec<T> {
+        if elem.is_zero() {
+            return Vec {
+                buf: RawVec::with_capacity_zeroed(n),
+                len: n,
+            }
+        }
+        let mut v = Vec::with_capacity(n);
+        v.extend_with(n, ExtendElement(elem));
+        v
+    }
+}
+
+unsafe trait IsZero {
+    /// Whether this value is zero
+    fn is_zero(&self) -> bool;
+}
+
+macro_rules! impl_is_zero {
     ($t: ty, $is_zero: expr) => {
-        impl SpecFromElem for $t {
+        unsafe impl IsZero for $t {
             #[inline]
-            fn from_elem(elem: $t, n: usize) -> Vec<$t> {
-                if $is_zero(elem) {
-                    return Vec {
-                        buf: RawVec::with_capacity_zeroed(n),
-                        len: n,
-                    }
-                }
-                let mut v = Vec::with_capacity(n);
-                v.extend_with(n, ExtendElement(elem));
-                v
+            fn is_zero(&self) -> bool {
+                $is_zero(*self)
             }
         }
-    };
+    }
 }
 
-impl_spec_from_elem!(i8, |x| x == 0);
-impl_spec_from_elem!(i16, |x| x == 0);
-impl_spec_from_elem!(i32, |x| x == 0);
-impl_spec_from_elem!(i64, |x| x == 0);
-impl_spec_from_elem!(i128, |x| x == 0);
-impl_spec_from_elem!(isize, |x| x == 0);
+impl_is_zero!(i8, |x| x == 0);
+impl_is_zero!(i16, |x| x == 0);
+impl_is_zero!(i32, |x| x == 0);
+impl_is_zero!(i64, |x| x == 0);
+impl_is_zero!(i128, |x| x == 0);
+impl_is_zero!(isize, |x| x == 0);
+
+impl_is_zero!(u16, |x| x == 0);
+impl_is_zero!(u32, |x| x == 0);
+impl_is_zero!(u64, |x| x == 0);
+impl_is_zero!(u128, |x| x == 0);
+impl_is_zero!(usize, |x| x == 0);
+
+impl_is_zero!(f32, |x: f32| x.to_bits() == 0);
+impl_is_zero!(f64, |x: f64| x.to_bits() == 0);
+
+unsafe impl<T: ?Sized> IsZero for *const T {
+    #[inline]
+    fn is_zero(&self) -> bool {
+        (*self).is_null()
+    }
+}
 
-impl_spec_from_elem!(u16, |x| x == 0);
-impl_spec_from_elem!(u32, |x| x == 0);
-impl_spec_from_elem!(u64, |x| x == 0);
-impl_spec_from_elem!(u128, |x| x == 0);
-impl_spec_from_elem!(usize, |x| x == 0);
+unsafe impl<T: ?Sized> IsZero for *mut T {
+    #[inline]
+    fn is_zero(&self) -> bool {
+        (*self).is_null()
+    }
+}
 
-impl_spec_from_elem!(f32, |x: f32| x.to_bits() == 0);
-impl_spec_from_elem!(f64, |x: f64| x.to_bits() == 0);
+unsafe impl<T> IsZero for Option<T> {
+    #[inline]
+    fn is_zero(&self) -> bool {
+        if self.is_none() {
+            // Not all Option<T>'s None are represented with zeroes, so check that
+            // a zeroed Option<T> is, indeed, None.
+            // The compiler should be able to figure all this out statically at
+            // compile time.
+            let a: Option<T> = unsafe { mem::zeroed() };
```

Comment:
This seems like undefined behavior.

Reply:
It's not entirely clear to me from the std::mem::zeroed doc that the way it's (ab)used here is actually UB. The value is explicitly not drop to avoid any definite UB, and is_none() is going to look at the discriminant, which will be zero, and return whether that value corresponds to the discriminant for None or not. And if None is not backed by a zero discriminant, it's because the compiler decided that zero was an interesting valid value for Some(something).

Reply:
Creating a value that can't exist is inherently UB.

Reply:
How can it not exist? None is either zero, or some other value because the compiler decided to allocate zero to Some(something). Conversely, that means zero is a valid value, but not necessarily None.

Reply:
Well, yes, arguably, zero for places other than the discriminant /might/ be invalid. But we're only even looking at the discriminant.

Reply:
`a.is_none()` looks at the discriminant of `a`.

Reply:
Would you prefer something like:
```rust
    if a.is_none() {
        let b: Option<T> = unsafe { std::mem::zeroed() };
        let result = std::mem::discriminant(&a) == std::mem::discriminant(&b);
        std::mem::forget(b);
        result
    } else {
        false
    }
```
which generates the same code.

Reply:
Maybe should expand the unsafe block to make it clearer.

Reply:
That's not a realistic example, since Option<&u32>'s None *is* 0, and fits in the same size as a pointer because &u32 itself is never 0, so the discriminant for Some() is non-zero. AFAIK, the only reason the compiler would choose a non 0 discriminant for None is if, like for Option<char>, there's room for invalid T values in the bit representation of T, *and* 0 is a valid value of T.

Reply:
`&u32` is never 1 either. The compiler could choose any value that isn't a multiple of 4.

Reply:
If it did, that would mean statics initialized to None would not end up in `.bss`.

Reply:
That's completely irrelevant to the soundness of this code.